android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/Utils.java

```diff
@@ -57,82 +57,170 @@
-    String apiKey = getApiKey(context);
-    String pwsEndpoint = getCurrentPwsEndpointUrl(context);
-    int apiVersion = getCurrentPwsEndpointVersion(context);
-    // Check for case where last saved settings was the Google PWS, but user has reinstalled
-    // the app without the API key.
-    if (apiVersion >= 2 && apiKey.isEmpty()){
-      pwsEndpoint = PROD_ENDPOINT;
-      apiVersion = PROD_ENDPOINT_VERSION;
-      Toast.makeText(context, R.string.error_api_key_no_longer_available,
-                     Toast.LENGTH_SHORT).show();
+  private static int readCustomPwsEndpointVersion(Context context) {
+    SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);
+    return Integer.parseInt(sharedPref.getString(
+      context.getString(R.string.custom_pws_version_key),
+      context.getString(R.string.custom_pws_version_default)));
+  }
+
+  private static String readCustomPwsEndpointApiKey(Context context) {
+    SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);
+    return sharedPref.getString(context.getString(R.string.custom_pws_api_key_key), "");
+  }
+
+
+
+  public static String convertToSettingsEndpointFormat(String pwsUrl, int pwsVersion,
+                                                       String apiKey) {
+    return pwsUrl + SEPARATOR + pwsVersion + SEPARATOR + apiKey;
+  }
+
+  public static String getDefaultPwsEndpointPreferenceString(Context context) {
+    if (isGoogleApiKeyAvailable(context)) {
+      return convertToSettingsEndpointFormat(GOOGLE_ENDPOINT, GOOGLE_ENDPOINT_VERSION,
+                                                        getGoogleApiKey(context));
     }
-    physicalWebCollection.setPwsEndpoint(pwsEndpoint, apiVersion, apiKey);
+    return convertToSettingsEndpointFormat(PROD_ENDPOINT, PROD_ENDPOINT_VERSION, "");
+  }
+
+  public static void setPwsEndpointPreference(Context context, String endpoint) {
+    SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);
+    SharedPreferences.Editor editor = sharedPref.edit();
+    editor.putString(context.getString(R.string.pws_endpoint_setting_key),
+                     endpoint).commit();
+  }
+
+  public static void setSharedPreferencesDefaultValues(Context context) {
+    PreferenceManager.setDefaultValues(context, R.xml.settings, false);
+    setPwsEndpointPreference(context, getCurrentPwsEndpointString(context));
+  }
+
+  public static boolean setPwsEndpoint(Context context, PwsClient pwsClient) {
+    PwsEndpoint endpoint = getCurrentPwsEndpoint(context);
+    pwsClient.setEndpoint(endpoint.url, endpoint.apiVersion, endpoint.apiKey);
+    return endpoint.neededApiKey;
   }
 
-  public static void setPwsEndPointToGoogle(Context context, PwsClient pwsClient){
-    pwsClient.setEndpoint(GOOGLE_ENDPOINT, GOOGLE_ENDPOINT_VERSION, getApiKey(context));
+  public static boolean setPwsEndpoint(Context context,
```

Comment:
Document the return values on these...I think it'd make more sense to return whether it was successful (the opposite of what you are returning).